Update docs (#1)
cycloid-community-catalog/stack-lemp: README.md

> # stack-lemp (Linux, Nginx, MySQL, PHP)

Reviewer: ```suggestion
# stack-lemp (Linux, NGINX, MySQL, PHP)
```

This stack will create A full LEMP infrastructure based on ASG and RDS database.

  * Linux operating system

>   * Nginx (Pronounced as Engine-X) web server

Reviewer: ```suggestion
  * NGINX (Pronounced as Engine-X) web server
```

  * MySQL (RDS) database server

>   * PHP-fpm for dynamic data processing

Reviewer: ```suggestion
  * PHP-FPM for dynamic data processing
```

# Architecture

<p align="center">
<img src="docs/diagram.png" width="400">
</p>

>   * **ALB**: Amazon application loadbalancer

Reviewer: ```suggestion
  * **ALB**: Amazon Application Load Balancer
```

>   * **ASG**: Autoscaling group for fronts

Reviewer: ```suggestion
  * **ASG**: Amazon Auto Scaling group for fronts
```

>   * **Elasticache** (optional): Amazon Elasticache (redis)

Reviewer: ```suggestion
  * **ElastiCache** (optional): Amazon ElastiCache (Redis)
```

  * **S3 bucket** (optional): public medias bucket

# Requirements

In order to run this task, couple elements are required within the infrastructure:

  * Having a VPC with private & public subnets [Here](https://docs.aws.amazon.com/vpc/latest/userguide/getting-started-ipv4.html#getting-started-create-vpc)
  * Having a bastion server to run Ansible like described [Here](https://docs.cycloid.io/advanced-guide/ansible-integration.html#standard-usage)
  * Having an S3 bucket to store Terraform remote states [Here](https://docs.aws.amazon.com/quickstarts/latest/s3backup/step-1-create-bucket.html)
  * Having an S3 bucket for website code **WITH** versioning enable [Here](https://docs.aws.amazon.com/quickstarts/latest/s3backup/step-1-create-bucket.html)

# Details

## Pipeline

> **Note** The pipeline contains a manual approval between terraform plan and terraform apply.
> That means if you trigger a terraform plan, to apply it, you have to go on terraform apply job
> and click on the `+` button to trigger it.

<img src="docs/pipeline.png" width="800">

**Jobs description**

>   * `build-worker-ami` : Build Front Amazon image AMI using Packer and Ansible.

Reviewer 2: build-ami-front* it's the name of the job, I guess it's a wrong copy/paste

  * `build-application`: Runs the appropriate php/composer commands and build a release file of the code to put it on S3.

>   * `unittest`: Dummy job meant to eventually be replaced by proper tests or removed.

Reviewer 2: unittests* according to the pipeline screenshot

>   * `terraform-plan`: Terraform job that will simply make a plan of the stack.

Reviewer 2: I'd put the `terraform-plan` before the `unittests` one as it is at the same level as the `buildapplication` pipeline wise, WDYT?

  * `terraform-apply`: Terraform job similar to the plan one, but will actually create/update everything that needs to. Please see the plan diff for a better understanding.
  * `application-deployment`: Simply trigger a deployment using Ansible of the last version of the code on existing instances.
  * `functional-tests`: Dummy job meant to eventually be replaced by proper functional or removed.

>   * `terraform-destroy`: :warning: Terraform job meant to destroy the whole stack - **NO CONFIRMATION ASKED**. If triggered, the full project **WILL** be destroyed. Use with caution.

Reviewer 2: Shall we specify that it's not displayed in the pipeline screenshot, but is on a dedicated pipeline tab?

Author: I think that's ok if this job is little bit harder to found :p

**Outputs**

| Name | Description |
|------|-------------|
| alb_front_dns_name | DNS name of the front alb. |
| alb_front_zone_id | Zone ID of the front alb. |
| elasticache_address | Address of the elasticache. |
| iam_s3-medias_user_key | Access key of the dedicated IAM user to access to the media S3 bucket. |
| iam_s3-medias_user_secret | Access secret key of the dedicated IAM user to access to the media S3 bucket. |
| rds_address | Address of the RDS database. |
| rds_database | Database name of the RDS database. |
| rds_port | Port of the RDS database. |
| rds_username | Username of the RDS database. |
| s3_medias | S3 bucket name dedicated to medias. |

## Ansible

  * Playbook and packer config to build a debian image with telegraf, fluentd, nginx and php-fpm installed

|Name|Description|Type|Default|Required|
|---|---|:---:|:---:|:---:|
|`cycloid_files_watched`|Provide log files you want to export to Cycloid logs.|`-`|`<Default log files watched>`|`False`|
|`nginx_sites`|Contain Nginx vhosts to create on front servers. A default application and metrics vhosts are already provided.|`dict`|`<metric and application vhost>`|`False`|

> |`php_version_to_install`|PHP fpm version to install.|`-`|`7.2`|`False`|

Reviewer 2: I assume all the tables of params for pipeline, terraform/ansible/ouptut was generated by your script? Or shall I check it too?

Author: yes tabs are generated by my scripts. From comments in the code

# Molecule tests

> Requires a bucket which contains a build of magento sources and AWS accesskey

Reviewer: ```suggestion
Requires a bucket which contains a build of magento sources and AWS access key
```
